aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h

```diff
@@ -187,12 +193,22 @@ namespace impl {
     static_assert(guts::false_t<T>::value, "You tried to register a kernel with an unsupported integral output type. Please use int64_t instead.");
   };
 
+  template<class T, bool AllowDeprecatedTypes>
+  struct ivalue_to_arg final {
+    static T call(IValue&& v) {
+      assert_is_valid_input_type<T, AllowDeprecatedTypes>();
+      return std::move(v).to<T>();
+    }
+  };
 
   template<class T, bool AllowDeprecatedTypes>
-  T ivalue_to_arg(IValue&& v) {
-    assert_is_valid_input_type<T, AllowDeprecatedTypes>();
-    return std::move(v).to<T>();
-  }
+  struct ivalue_to_arg<ArrayRef<T>, AllowDeprecatedTypes> final {
+    // If an argument is ArrayRef<T>, convert the IValue to a std::vector<T> and pass that
+    // to the operator. std::vector<T> is implicitly convertible to ArrayRef<T>.
+    static std::vector<T> call(IValue&& v) {
+      return ivalue_to_arg<std::vector<T>, AllowDeprecatedTypes>::call(std::move(v));
```

Comment:
This sounds really slow. cc @zdevito (I guess this is a consequence of uniform List representation meaning that conversion to an ArrayRef is not O(1)? That sucks!)

Reply:
yes that's a consequence of that. In light of that, I think we should consider only use `ArrayRef` as a convenience feature for kernels that need the memory layout of an `ArrayRef`, for example for passing a pointer to an `ArrayRef` with strides to a CUDA kernel, but everything else should use `torch::List` which avoids the copy.

Reply:
@smessmer I think that's a pretty big step to take, so I'd like to see more evidence for this recommendation. For example, the C++ API takes ArrayRef as a public API. I can't see how ArrayRef O(1) converts to torch::List.